stan/math/opencl/prim/add_diag.hpp

```diff
@@ -0,0 +1,41 @@
+#ifndef STAN_MATH_OPENCL_PRIM_ADD_DIAG_HPP
+#define STAN_MATH_OPENCL_PRIM_ADD_DIAG_HPP
+#ifdef STAN_OPENCL
+#include <stan/math/opencl/prim/size.hpp>
+#include <stan/math/opencl/kernel_generator.hpp>
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/err.hpp>
+
+namespace stan {
+namespace math {
+
+/**
+ * Returns a Matrix with values added along the main diagonal
+ *
+ * @tparam T_m type of input kernel generator expression for the input matrix
+ * @tparam T_a type of input kernel generator expression to add along the
+ * diagonal
+ *
+ * @param mat input kernel generator expression
+ * @param to_add scalar value or input kernel generator expression to add along
+ * the diagonal
+ * @return a kernel generator expressio with to_add added along main diagonal
+ */
+template <typename T_m, typename T_a,
+          require_all_kernel_expressions_and_none_scalar_t<T_m>* = nullptr,
+          require_all_kernel_expressions_t<T_a>* = nullptr>
+inline auto add_diag(T_m&& mat, T_a&& to_add) {  // NOLINT
```

Comment:
The way this is implemented it modifies the mat in place. This should be either documented or you should make a copy before changing it.